### ISSUE TITLE
UnboundLocalError: local variable 'tn' referenced before assignment

### ISSUE THREAD

**Issue by gadilashashank** [nicfit/eyeD3#327]
So I was using eyeD3 ( python ) to set ID3 tags to mp3 files for a personal project when the given traceback occurred in ```eyed3/id3/tag.py``` As a temporary hack I changed _setNum(self, fid, val) on line 320 in the said file as follows:

```python3  
def _setNum(self, fid, val):
        tn, tt = None, None # Added this line 
        if type(val) is tuple:
            tn, tt = val
        elif type(val) is int:
            tn, tt = val, None
        elif val is None:
            tn, tt = None, None

        n = (tn, tt)
```
Although the UnboundLocalError was gone but the track number tag was not being assigned properly (I didn't expect to either) hence this issue for hopefully a proper fix (or guidance so that I can fix it myself and open a PR)

The error occurred in a Py3.7 virtual environment

```
Python 3.7.2 (default, Jan 10 2019, 23:51:51)
[GCC 8.2.1 20181127] on linux
```
 
```
$ eyeD3 --version
eyed3.plugins:WARNING: Plugin '('lastfm.py', '/home/shashank/Projects/CompleteMusic-2.0/lib/python3.7/site-packages/eyed3/plugins')' requires packages that are not installed: cannot import name 'COVER_EXTRA_LARGE' from 'pylast' (/home/shashank/Projects/CompleteMusic-2.0/lib/python3.7/site-packages/pylast/__init__.py)
0.8.10
```

```
Traceback (most recent call last):
File "mio.py", line 182, in <module>
test_object = MusicIndiaOnline("Balam Pichkari")
File "mio.py", line 31, in __init__
self.download_song()
File "mio.py", line 77, in download_song
audiofile.tag.track_num = self.metadata['track-number']
File "[I have redacted this]/lib/python3.7/site-packages/eyed3/core.py", line 173, in track_num
self._setTrackNum(v)
File "[I have redacted this]/lib/python3.7/site-packages/eyed3/id3/tag.py", line 303, in _setTrackNum
self._setNum(frames.TRACKNUM_FID, val)
File "[I have redacted this]/lib/python3.7/site-packages/eyed3/id3/tag.py", line 328, in _setNum
n = (tn, tt)
UnboundLocalError: local variable 'tn' referenced before assignment
```

I hope I have given enough info. I'd also like to clarify that apart from the above change in source code I haven't changed anything else.

**Comment by nicfit** [comment]
Thanks for the report @gabrieldiego, I've made the change for the next eyeD3 release.

**Comment by nicfit** [comment]
@gadilashashank  Do you happen to know what `val` was passed in to the _setNum function? I'm thinking more should be done in the patch. Currently, invalid values are silently ignored.

**Comment by gabrieldiego** [comment]
@nicfit Was this message directed to me? I guess you wanted to refer to @gadilashashank , right?

**Comment by nicfit** [comment]
Yea, it was. I've made the edit. Thanks.

**Comment by gadilashashank** [comment]
@nicfit  Sorry I've been gone for a while. I have printed different local variables and their respective outputs. I hope it makes some sense for you. The track number of the particular song I have used is set at 3 which is printed in some instances.

```
    def _setNum(self, fid, val):
        if type(val) is tuple:
            tn, tt = val
        elif type(val) is int:
            tn, tt = val, None
        elif val is None:
            tn, tt = None, None

        n = (tn, tt)
---
Traceback (most recent call last):
  [Some traceback which leads to]
  File "[]/lib/python3.7/site-packages/eyed3/core.py", line 173, in track_num
    self._setTrackNum(v)
  File "[]/lib/python3.7/site-packages/eyed3/id3/tag.py", line 303, in _setTrackNum
    self._setNum(frames.TRACKNUM_FID, val)
  File "[]/lib/python3.7/site-packages/eyed3/id3/tag.py", line 330, in _setNum
    n = (tn, tt)
UnboundLocalError: local variable 'tn' referenced before assignment
```

```
    def _setNum(self, fid, val):
         print(val)
         print("SOME DELIMITER")
         print(fid)
        if type(val) is tuple:
            tn, tt = val
        elif type(val) is int:
            tn, tt = val, None
        elif val is None:
            tn, tt = None, None

        n = (tn, tt)
---
None
SOME DELIMITER
b'TRCK'
None
SOME DELIMITER
b'TRCK'
3
SOME DELIMITER
b'TRCK'
Traceback (most recent call last):
  [Some traceback which leads to]
   audiofile.tag.track_num = self.metadata['track-number']
  File "[]/lib/python3.7/site-packages/eyed3/core.py", line 173, in track_num
    self._setTrackNum(v)
  File "[]/lib/python3.7/site-packages/eyed3/id3/tag.py", line 303, in _setTrackNum
    self._setNum(frames.TRACKNUM_FID, val)
  File "[]/lib/python3.7/site-packages/eyed3/id3/tag.py", line 332, in _setNum
    n = (tn, tt)
UnboundLocalError: local variable 'tn' referenced before assignment

```
```
    def _setNum(self, fid, val):
        tn, tt = None, None
        print( val)
        print("SOME DELIMITER")
        print(fid)
        if type(val) is tuple:
            tn, tt = val
        elif type(val) is int:
            tn, tt = val, None
        elif val is None:
            tn, tt = None, None

        n = (tn, tt)
---
None
SOME DELIMITER
b'TRCK'
None
SOME DELIMITER
b'TRCK'
3
SOME DELIMITER
b'TRCK'
None
SOME DELIMITER
b'TRCK'
```

I have only included traceback which calls code related to eyeD3 and left out the rest, and omitted the exact paths due to privacy reasons. I don't think these are necessary/related to debugging this issue.

**Comment by gadilashashank** [comment]
More variations; I honestly don't know why I did this.
```
    def _setNum(self, fid, val):
        print(tt)
        if type(val) is tuple:
            tn, tt = val
        elif type(val) is int:
            tn, tt = val, None
        elif val is None:
            tn, tt = None, None

        n = (tn, tt)
---
Traceback (most recent call last):
    [Some omitted traceback]
    audiofile = eyed3.load(self.metadata['name'] + ".mp3")
  File "[]/lib/python3.7/site-packages/eyed3/core.py", line 79, in load
    return mp3.Mp3AudioFile(path, tag_version)
  File "[]/lib/python3.7/site-packages/eyed3/mp3/__init__.py", line 172, in __init__
    core.AudioFile.__init__(self, path)
  File "[]/lib/python3.7/site-packages/eyed3/core.py", line 257, in __init__
    self._read()
  File "[]/site-packages/eyed3/mp3/__init__.py", line 177, in _read
    self._tag = id3.Tag()
  File "[]/lib/python3.7/site-packages/eyed3/id3/tag.py", line 55, in __init__
    core.Tag.__init__(self, **kwargs)
  File "/[]/site-packages/eyed3/core.py", line 181, in __init__
    self.track_num = track_num
  File "[]/lib/python3.7/site-packages/eyed3/core.py", line 173, in track_num
    self._setTrackNum(v)
  File "[]/lib/python3.7/site-packages/eyed3/id3/tag.py", line 303, in _setTrackNum
    self._setNum(frames.TRACKNUM_FID, val)
  File "[]/lib/python3.7/site-packages/eyed3/id3/tag.py", line 321, in _setNum
    print(tt)
UnboundLocalError: local variable 'tt' referenced before assignment
```

**Comment by gadilashashank** [comment]
Oh wow. I figured out what's happening and **it's totally my fault**. 
Here's the snippet of my code where the I was doing wrong.
```python3
audiofile = eyed3.load("Some-song.mp3")
audiofile.initTag()
audiofile.tag.track_num = **SOME_PARAMETER**
audiofile.tag.save()
```
Turns out **SOME_PARAMETER** that I was sending to the function this snippet was in is of type "str" whereas eyeD3 expects it to be of type "int". The moment I changed the snippet to 
```python3
audiofile = eyed3.load("Some-song.mp3")
audiofile.initTag()
audiofile.tag.track_num = int(**SOME_PARAMETER**)
audiofile.tag.save()
```
and with no changes in eyeD3 code base everything works perfectly and as expected.

I was in the wrong by sending track number of the wrong type. So the fix would be to check if the correct type was sent in (which is already committed in the latest). So I guess this issue can be closed and **I'm sorry for all the trouble**. 
Also my previous comment is not so useful given I have identified the mistake I was making.

**Comment by nicfit** [comment]
No worries, `_setNum` became better from this discussion. Now, you'll get a ValueError with a more helpful message.  https://github.com/nicfit/eyeD3/blob/master/src/eyed3/id3/tag.py#L327

**Comment by nicfit** [comment]
Keeping this open because I think `TypeError` should be raised instead of `ValueError`, and taking it even further, a `ValueError` should be raised when a negative int, or non 2-tuple, or tuple containing bogus values, etc. is passed in.

**Comment by nicfit** [comment]
Fixed in de76fa8cb67b0a60167ed49449c142bc31896560
